In [1]:
#importing the basic libraries
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import pairwise_distances
from scipy.spatial.distance import cosine,correlation

In C:\Users\KUMAR SANU\anaconda3\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle: 
The text.latex.preview rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.
In C:\Users\KUMAR SANU\anaconda3\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle: 
The mathtext.fallback_to_cm rcparam was deprecated in Matplotlib 3.3 and will be removed two minor releases later.
In C:\Users\KUMAR SANU\anaconda3\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle: Support for setting the 'mathtext.fallback_to_cm' rcParam is deprecated since 3.3 and will be removed two minor releases later; use 'mathtext.fallback : 'cm' instead.
In C:\Users\KUMAR SANU\anaconda3\lib\site-packages\matplotlib\mpl-data\stylelib\_classic_test.mplstyle: 
The validate_bool_maybe_none function was deprecated in Matplotlib 3.3 and will be removed two minor releases later.
In C:\Users\KUMAR SANU\anaconda3\lib\site-packages\matplotlib\mpl-data\s

In [3]:
ls

 Volume in drive C is Windows-SSD
 Volume Serial Number is 6E82-34BE

 Directory of C:\Users\KUMAR SANU\Assignment\Recommendation system

18-03-2021  12:17    <DIR>          .
18-03-2021  12:17    <DIR>          ..
18-03-2021  12:17    <DIR>          .ipynb_checkpoints
18-03-2021  12:17           133,382 001122 Recommendation System.ipynb
02-03-2021  15:06           498,908 book.csv
18-03-2021  10:59           117,775 Recommendation.ipynb
01-03-2021  10:42                66 RS.txt
               4 File(s)        750,131 bytes
               3 Dir(s)  182,445,891,584 bytes free


In [4]:
#reading the dataset
books=pd.read_csv("book.csv",encoding='cp1252')
books.head()

sn  userid                                             tittle  rating
0   1  276726                                Classical Mythology       5
1   2  276729                                       Clara Callan       3
2   3  276729                               Decision in Normandy       6
3   4  276736  Flu: The Story of the Great Influenza Pandemic...       8
4   5  276737                             The Mummies of Urumchi       6

In [5]:
data=books.iloc[:,1:]
data

userid                                             tittle  rating
0     276726                                Classical Mythology       5
1     276729                                       Clara Callan       3
2     276729                               Decision in Normandy       6
3     276736  Flu: The Story of the Great Influenza Pandemic...       8
4     276737                             The Mummies of Urumchi       6
...      ...                                                ...     ...
9995  162121       American Fried: Adventures of a Happy Eater.       7
9996  162121                              Cannibal In Manhattan       9
9997  162121                    How to Flirt: A Practical Guide       7
9998  162121                                           Twilight       8
9999  162129                       Kids Say the Darndest Things       6

[10000 rows x 3 columns]

## EDA

In [6]:
data.isnull().sum()

userid    0
tittle    0
rating    0
dtype: int64

In [8]:
df=data.pivot_table(values='rating', index="userid", columns="tittle",fill_value=0)
df

tittle   Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
userid                                                                     
8                          0                                           0   
9                          0                                           0   
10                         0                                           0   
12                         0                                           0   
14                         0                                           0   
...                      ...                                         ...   
278846                     0                                           0   
278849                     0                                           0   
278851                     0                                           0   
278852                     0                                           0   
278854                     0                                           0   

tittle   Repairing PC Drives &amp  '48  \
userid                                   
8                               0    0   
9                               0    0   
10                              0    0   
12                              0    0   
14                              0    0   
...                           ...  ...   
278846                          0    0   
278849                          0    0   
278851                          0    0   
278852                          0    0   
278854                          0    0   

tittle  'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
userid                                                                          
8                                                       0                       
9                                                       0                       
10                                                      0                       
12                                                      0                       
14                                                      0                       
...                                                   ...                       
278846                                                  0                       
278849                                                  0                       
278851                                                  0                       
278852                                                  0                       
278854                                                  0                       

tittle  ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
userid                                                                 
8                                                       0              
9                                                       0              
10                                                      0              
12                                                      0              
14                                                      0              
...                                                   ...              
278846                                                  0              
278849                                                  0              
278851                                                  0              
278852                                                  0              
278854                                                  0              

tittle  01-01-00: A Novel of the Millennium  \
userid                                        
8                                         0   
9                                         0   
10                                        0   
12                                        0   
14                                        0   
...                                     ...   
278846                                    0   
278849                                    0   
278

###  Building the Recommendation System using Cosine simil

In [9]:
similar_users=1-pairwise_distances(df.values,metric='cosine')
similar_users

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

In [10]:
sim_df=pd.DataFrame(similar_users)
sim_df

0     1     2     3     4     5     6     7     8     9     ...  2172  \
0      1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
1      0.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2      0.0   0.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
3      0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
4      0.0   0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
2177   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2178   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2179   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2180   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   
2181   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   0.0   

      2173  2174  2175  2176  2177  2178  2179  2180  2181  
0      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
1      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
2      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
3      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
4      0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...  
2177   0.0   0.0   0.0   0.0   1.0   0.0   0.0   0.0   0.0  
2178   0.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0   0.0  
2179   0.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   0.0  
2180   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0  
2181   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   1.0  

[2182 rows x 2182 columns]

In [11]:
sim_df.index=df.index
sim_df.columns=df.index
sim_df

userid  8       9       10      12      14      16      17      19      \
userid                                                                   
8          1.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
9          0.0     1.0     0.0     0.0     0.0     0.0     0.0     0.0   
10         0.0     0.0     1.0     0.0     0.0     0.0     0.0     0.0   
12         0.0     0.0     0.0     1.0     0.0     0.0     0.0     0.0   
14         0.0     0.0     0.0     0.0     1.0     0.0     0.0     0.0   
...        ...     ...     ...     ...     ...     ...     ...     ...   
278846     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278849     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278851     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278852     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278854     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

userid  22      26      ...  278831  278832  278836  278843  278844  278846  \
userid                  ...                                                   
8          0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
9          0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
10         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
12         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
14         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
...        ...     ...  ...     ...     ...     ...     ...     ...     ...   
278846     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     1.0   
278849     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278851     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278852     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278854     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

userid  278849  278851  278852  278854  
userid                                  
8          0.0     0.0     0.0     0.0  
9          0.0     0.0     0.0     0.0  
10         0.0     0.0     0.0     0.0  
12         0.0     0.0     0.0     0.0  
14         0.0     0.0     0.0     0.0  
...        ...     ...     ...     ...  
278846     0.0     0.0     0.0     0.0  
278849     1.0     0.0     0.0     0.0  
278851     0.0     1.0     0.0     0.0  
278852     0.0     0.0     1.0     0.0  
278854     0.0     0.0     0.0     1.0  

[2182 rows x 2182 columns]

In [12]:
np.fill_diagonal(similar_users,0)
sim_df

userid  8       9       10      12      14      16      17      19      \
userid                                                                   
8          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
9          0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
10         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
12         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
14         0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
...        ...     ...     ...     ...     ...     ...     ...     ...   
278846     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278849     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278851     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278852     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278854     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

userid  22      26      ...  278831  278832  278836  278843  278844  278846  \
userid                  ...                                                   
8          0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
9          0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
10         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
12         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
14         0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
...        ...     ...  ...     ...     ...     ...     ...     ...     ...   
278846     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278849     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278851     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278852     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278854     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

userid  278849  278851  278852  278854  
userid                                  
8          0.0     0.0     0.0     0.0  
9          0.0     0.0     0.0     0.0  
10         0.0     0.0     0.0     0.0  
12         0.0     0.0     0.0     0.0  
14         0.0     0.0     0.0     0.0  
...        ...     ...     ...     ...  
278846     0.0     0.0     0.0     0.0  
278849     0.0     0.0     0.0     0.0  
278851     0.0     0.0     0.0     0.0  
278852     0.0     0.0     0.0     0.0  
278854     0.0     0.0     0.0     0.0  

[2182 rows x 2182 columns]

In [13]:
sim_df=sim_df.replace(0,np.nan)
sim_df

userid  8       9       10      12      14      16      17      19      \
userid                                                                   
8          NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
9          NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
10         NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
12         NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
14         NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
...        ...     ...     ...     ...     ...     ...     ...     ...   
278846     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
278849     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
278851     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
278852     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
278854     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

userid  22      26      ...  278831  278832  278836  278843  278844  278846  \
userid                  ...                                                   
8          NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
9          NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
10         NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
12         NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
14         NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
...        ...     ...  ...     ...     ...     ...     ...     ...     ...   
278846     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
278849     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
278851     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
278852     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
278854     NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   

userid  278849  278851  278852  278854  
userid                                  
8          NaN     NaN     NaN     NaN  
9          NaN     NaN     NaN     NaN  
10         NaN     NaN     NaN     NaN  
12         NaN     NaN     NaN     NaN  
14         NaN     NaN     NaN     NaN  
...        ...     ...     ...     ...  
278846     NaN     NaN     NaN     NaN  
278849     NaN     NaN     NaN     NaN  
278851     NaN     NaN     NaN     NaN  
278852     NaN     NaN     NaN     NaN  
278854     NaN     NaN     NaN     NaN  

[2182 rows x 2182 columns]

In [14]:
similarity=pd.DataFrame(sim_df.idxmax(axis=1,skipna=True))
similarity

0
userid          
8            NaN
9            NaN
10           NaN
12           NaN
14           NaN
...          ...
278846       NaN
278849       NaN
278851  278202.0
278852       NaN
278854       NaN

[2182 rows x 1 columns]

### Users with maximum similarity

In [15]:
similarity.dropna()

0
userid          
19      278418.0
51        3757.0
53        1996.0
82         882.0
83      276861.0
...          ...
278694  161831.0
278715  277710.0
278843  277959.0
278844  276813.0
278851  278202.0

[368 rows x 1 columns]

### Checking if the model works correctly

In [17]:
data[(data['userid']==278851) | (data['userid']==278202)]

userid                                             tittle  rating
1706  278202  The Florabama Ladies' Auxiliary &amp; Sewing C...       3
1707  278202                Howl and Other Poems (Pocket Poets)       2
1708  278202  Frankenstein: Complete, Authoritative Text Wit...       4
1709  278202  Short Stories of Ernest Hemingway (A Scribner ...       5
1710  278202  Back When We Were Grownups : A Novel (Ballanti...       3
1711  278202                              To Kill a Mockingbird       9
1712  278202                         Birds Bees And Babies 1994       3
1713  278202                          The Cat Who Played Brahms      10
1714  278202                                       Black Notice       3
1715  278202  Irish Eyes: A Nuala Anne McGrail Novel (Nuala ...       5
1716  278202  Totally Useless Office Skills: 75 Great Ways t...       3
1717  278202                   Catechism of the Catholic Church       1
1718  278202  Father McBride's Family Catechism: Based on th...       8
1719  278202                         The Missionary's Catechism       3
1720  278202                The Enthusiasms of Roberston Davies       4
1721  278202  As Nature Made Him: The Boy Who Was Raised As ...       4
1722  278202                Something Under the Bed Is Drooling      10
1723  278202  The War Against Population: The Economics and ...       7
1724  278202                                  Mere Christianity       3
1725  278202                      Haroun and the Sea of Stories       8
1726  278202  Wedding Day Murder: A Lucy Stone Mystery (Meie...       4
1727  278202         One for the Money (A Stephanie Plum Novel)       4
2378  278851               So You Want to Be a Stay-At-Home Mom       5
2379  278851                                       Black Notice       5
2380  278851                                       Mystic River       5
2381  278851                                            Hitched       7
2382  278851                              Fantastic Imagination       5
2383  278851                                      Six of Swords       5
2384  278851                                      Paradise Wild       7
2385  278851                            iI Paradiso Degli Orchi       7
2386  278851     The Celestine Prophecy : An Experiential Guide       5
2387  278851  Huis Clos Suivi de Les Mouches (Folio Ser. No....       7
2388  278851                              Escape from Kathmandu       8
2389  278851                                   McNally's Chance       8
2390  278851         The Year's Best Horror Stories: Series XIV      10
2391  278851                            Pay It Forward: A Novel       7

##### We can see that both users that our model predicted to be similar have given rating to the book titled Black Notice. Therefore, we can conclude that the model is finding a relation between the users using the rating

In [18]:
user_1=data[data['userid']==278851]
user_2=data[data['userid']==278202]
pd.merge(user_1,user_2,on='tittle',how='outer')

userid_x                                             tittle  rating_x  \
0   278851.0               So You Want to Be a Stay-At-Home Mom       5.0   
1   278851.0                                       Black Notice       5.0   
2   278851.0                                       Mystic River       5.0   
3   278851.0                                            Hitched       7.0   
4   278851.0                              Fantastic Imagination       5.0   
5   278851.0                                      Six of Swords       5.0   
6   278851.0                                      Paradise Wild       7.0   
7   278851.0                            iI Paradiso Degli Orchi       7.0   
8   278851.0     The Celestine Prophecy : An Experiential Guide       5.0   
9   278851.0  Huis Clos Suivi de Les Mouches (Folio Ser. No....       7.0   
10  278851.0                              Escape from Kathmandu       8.0   
11  278851.0                                   McNally's Chance       8.0   
12  278851.0         The Year's Best Horror Stories: Series XIV      10.0   
13  278851.0                            Pay It Forward: A Novel       7.0   
14       NaN  The Florabama Ladies' Auxiliary &amp; Sewing C...       NaN   
15       NaN                Howl and Other Poems (Pocket Poets)       NaN   
16       NaN  Frankenstein: Complete, Authoritative Text Wit...       NaN   
17       NaN  Short Stories of Ernest Hemingway (A Scribner ...       NaN   
18       NaN  Back When We Were Grownups : A Novel (Ballanti...       NaN   
19       NaN                              To Kill a Mockingbird       NaN   
20       NaN                         Birds Bees And Babies 1994       NaN   
21       NaN                          The Cat Who Played Brahms       NaN   
22       NaN  Irish Eyes: A Nuala Anne McGrail Novel (Nuala ...       NaN   
23       NaN  Totally Useless Office Skills: 75 Great Ways t...       NaN   
24       NaN                   Catechism of the Catholic Church       NaN   
25       NaN  Father McBride's Family Catechism: Based on th...       NaN   
26       NaN                         The Missionary's Catechism       NaN   
27       NaN                The Enthusiasms of Roberston Davies       NaN   
28       NaN  As Nature Made Him: The Boy Who Was Raised As ...       NaN   
29       NaN                Something Under the Bed Is Drooling       NaN   
30       NaN  The War Against Population: The Economics and ...       NaN   
31       NaN                                  Mere Christianity       NaN   
32       NaN                      Haroun and the Sea of Stories       NaN   
33       NaN  Wedding Day Murder: A Lucy Stone Mystery (Meie...       NaN   
34       NaN         One for the Money (A Stephanie Plum Novel)       NaN   

    userid_y  rating_y  
0        NaN       NaN  
1   278202.0       3.0  
2        NaN       NaN  
3        NaN       NaN  
4        NaN       NaN  
5        NaN       NaN  
6        NaN       NaN  
7        NaN       NaN  
8        NaN       NaN  
9        NaN       NaN  
10       NaN       NaN  
11       NaN       NaN  
12       NaN       NaN  
13       NaN       NaN  
14  278202.0       3.0  
15  278202.0       2.0  
16  278202.0       4.0  
17  278202.0       5.0  
18  278202.0       3.0  
19  278202.0       9.0  
20  278202.0       3.0  
21  278202.0      10.0  
22  278202.0       5.0  
23  278202.0       3.0  
24  278202.0       1.0  
25  278202.0       8.0  
26  278202.0       3.0  
27  278202.0       4.0  
28  278202.0       4.0  
29  278202.0      10.0  
30  278202.0       7.0  
31  278202.0       3.0  
32  278202.0       8.0  
33  278202.0       4.0  
34  278202.0       4.0

##### We can now recommend all the other books that these two users have read. This can be done for all 368 pairs of users found

### Finding the similarity of the book based on the rating

In [19]:
# Function to find simlarity with other books based on ratings
def book_similarity(data, book_title):
    rating = data[book_title]
    similar = data.corrwith(rating)
    corr_df = pd.DataFrame(similar, columns=['Correlation'])
    corr_df.dropna(inplace=True)
    print(f'---Top 5 similar books for book "{book_title}" are:')
    print(corr_df[(corr_df['Correlation'] > 0) & (corr_df.index != book_title)].sort_values(by='Correlation', ascending=False).head(5))
    print("____________________\n")

In [28]:
data['tittle'].value_counts().head(10)

Fahrenheit 451                                     5
Ender's Game (Ender Wiggins Saga (Paperback))      4
Charlie and the Chocolate Factory                  4
The Amber Spyglass (His Dark Materials, Book 3)    4
The Subtle Knife (His Dark Materials, Book 2)      4
Vanished                                           4
Stardust                                           4
Strangers                                          3
Life Is Elsewhere                                  3
My Side of the Mountain                            3
Name: tittle, dtype: int64

In [26]:
famous_books=data['tittle'].value_counts().head(10).keys()

#### Top 5 books similar to top 10 famous books

In [24]:
for books in famous_books:
    book_similarity(df,books)

---Top 5 similar books for book "Fahrenheit 451" are:
                                                    Correlation
tittle                                                         
The Day I Swapped My Dad for 2 Goldfish                0.596326
Marriage By Contract  (36 Hours) (Harlequin 36 ...     0.596326
For The Love Of Beau (Delta Justice) (Delta Jus...     0.596326
Julia (Circle Of Friends) (Harlequin Superroman...     0.596326
Silent Witness                                         0.596326
____________________

---Top 5 similar books for book "Ender's Game (Ender Wiggins Saga (Paperback))" are:
                                                 Correlation
tittle                                                      
The Brothers Karamazov (Vintage Classics)           0.626405
Song of Solomon (Oprah's Book Club (Paperback))     0.626405
Come Before Winter and Share My Hope                0.547992
Non-manipulative selling                            0.547992
The Bear and the Dragon  

#### We can now recommend the books that are correlated to the people buying any of the books in the dataset